### ISSUE TITLE
Why does Cuda NLLCriterion mandate consecutive labels starting at 0?

### ISSUE THREAD

**Issue by themightyoarfish** [pytorch/pytorch#15666]
This is more of a question, but maybe a bug I'm not sure. I created a CIFAR10 variant and the labels were 1..10 as opposed to 0..9 as the `torchvision` one. When trying to feed that data through a model and backpropagating with CrossEntropy, I get
```
/home/share/software/sources/pytorch/aten/src/THCUNN/ClassNLLCriterion.cu:105: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long
 *, Dtype *, int, int, int, int, long) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [1,0,0] Assertion `t >= 0 && t < n_classes` failed.
/home/share/software/sources/pytorch/aten/src/THCUNN/ClassNLLCriterion.cu:105: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long
 *, Dtype *, int, int, int, int, long) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [20,0,0] Assertion `t >= 0 && t < n_classes` failed.
/home/share/software/sources/pytorch/aten/src/THCUNN/ClassNLLCriterion.cu:105: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long
 *, Dtype *, int, int, int, int, long) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [21,0,0] Assertion `t >= 0 && t < n_classes` failed.
```
so it seems to enforce that the labels start at 0. Is this intended behaviour, and if yes – why is it necessary? Is there documentation on this somewhere? It seems like users should be informed about this when creating their own dataset.

**Comment by zou3519** [comment]
It is documented here: https://pytorch.org/docs/master/nn.html#torch.nn.NLLLoss that the target labels must be in the range `[0, num_classes)`.

It is necessary because NLLLoss indexes into the input tensor at indices equal to the target labels.
